File: ChatGPT/clojure/output/output_18b.clj
;; ChatGPT-generated solution will go here.
(ns output_18b
  (:require [clojure.string :as str]
            [clojure.set :as set]
            [clojure.data.priority-map :refer [priority-map]]))

(defn parse-coord [line]
  (let [[x y] (str/split line #",")]
    [(Integer/parseInt x) (Integer/parseInt y)]))

(defn parse-input [lines]
  (map parse-coord (map str/trim (filter seq lines))))

(defn neighbors [[x y]]
  [[(inc x) y] [(dec x) y] [x (inc y)] [x (dec y)]])

(defn in-bounds? [[x y]]
  (and (<= 0 x 70) (<= 0 y 70)))

(defn bfs [start goal blocked]
  (loop [q (priority-map start 0)
         visited #{start}]
    (if-let [[[pos cost] & rest] (seq q)]
      (if (= pos goal)
        true
        (let [next (for [n (neighbors pos)
                         :when (and (in-bounds? n)
                                    (not (blocked n))
                                    (not (visited n)))]
                     n)
              q' (reduce (fn [q n] (assoc q n (inc cost))) (dissoc q pos) next)
              visited' (into visited next)]
          (recur q' visited')))
      false)))

(defn part [lines]
  (let [coords (parse-input lines)]
    (loop [fallen #{} remaining coords]
      (let [next (first remaining)
            fallen' (conj fallen next)]
        (if (not (bfs [0 0] [70 70] fallen'))
          (str (first next) "," (second next))
          (recur fallen' (rest remaining)))))))

(defn -main [& args]
  (let [filename (first args)
        lines (str/split-lines (slurp filename))]
    (println (part lines))))

File: ChatGPT/trans_test/clojure_to_python/output/output_18b.py
import heapq

def parse_coord(line):
    x, y = map(int, line.strip().split(","))
    return (x, y)

def parse_input(lines):
    return [parse_coord(line) for line in lines if line.strip()]

def neighbors(x, y):
    return [(x + 1, y), (x - 1, y), (x, y + 1), (x, y - 1)]

def in_bounds(x, y):
    return 0 <= x <= 70 and 0 <= y <= 70

def bfs(start, goal, blocked):
    heap = [(0, start)]
    visited = {start}

    while heap:
        cost, (x, y) = heapq.heappop(heap)
        if (x, y) == goal:
            return True
        for nx, ny in neighbors(x, y):
            if in_bounds(nx, ny) and (nx, ny) not in blocked and (nx, ny) not in visited:
                visited.add((nx, ny))
                heapq.heappush(heap, (cost + 1, (nx, ny)))
    return False

def part(lines):
    coords = parse_input(lines)
    fallen = set()
    for x, y in coords:
        fallen.add((x, y))
        if not bfs((0, 0), (70, 70), fallen):
            return f"{x},{y}"
    return None

def main():
    with open('input.txt') as f:
        lines = f.read().splitlines()
    print(part(lines))

if __name__ == '__main__':
    main()
